python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -1158,6 +1160,176 @@ def f():
       .isNotNull();
   }
 
+  @Test
+  void inferLoopVarType() {
+    var root = inferTypes("""
+      def f():
+        l = [1,2,3]
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(0), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(ObjectType.class)
+      .extracting(PythonType::unwrappedType)
+      .isInstanceOf(ClassType.class)
+      .extracting(PythonType::name)
+      .isEqualTo("int");
+  }
+
+  @Test
+  void inferLoopOverLiteralVarType() {
+    var root = inferTypes("""
+      def f():
+        for i in [1,2,3]:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(0), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(ObjectType.class)
+      .extracting(PythonType::unwrappedType)
+      .isInstanceOf(ClassType.class)
+      .extracting(PythonType::name)
+      .isEqualTo("int");
+  }
+
+  @Test
+  void inferLoopOverLiteralVarTypeCallable() {
+    var root = inferTypes("""
+      def foo():
+        print("foo")
+        
+      def bar():
+        print("bar")
+        
+      def f():
+        l = [foo, bar]
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root.statements().statements().get(2), ExpressionStatement.class::isInstance)
+      .map(ExpressionStatement.class::cast)
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnionType.class);
+    var candidates = ((UnionType) iType).candidates();
+    Assertions.assertThat(candidates)
+      .allMatch(FunctionType.class::isInstance)
+      .extracting(PythonType::name)
+      .contains("foo", "bar");
```

Comment:
Here also you can probably retrieve the function types of `foo` and `bar` directly.